### DISCUSSION TITLE
Qt6

### DISCUSSION THREAD

**Discussion by mchamberland** [nrc-cnrc/EGSnrc#1146] [Ideas]
It feels like Qt5 support was added just yesterday, but it's already been 7 years. Qt5.15 will only be supported until 2025, as far as I can tell.

Are there any plans to migrate to Qt6 soon?

Also, at least on Windows, it wasn't immediately trivial to download and install Qt5, even for me, so I would favor transitioning to Qt6 sooner rather than later.

**Comment by crcrewso** [comment]
I've been thinking about that. I think the best strategy is to update to 5.15 with the deprecation warnings turned on and remove the deprecated calls. 
https://doc.qt.io/qt-6/qtcore5-index.html
https://doc.qt.io/qt-6/portingguide.html

**Reply by mchamberland** [reply]
Yes, I agree. I've been working on porting a Qt5 application to Qt6, and it's fairly straightforward.

**Reply by crcrewso** [reply]
@mchamberland Did you do the last port? I remember starting it but it was definitely improved by someone before the patch was accepted

**Reply by mchamberland** [reply]
No, I did not. It must have been @rtownson...?

**Reply by rtownson** [reply]
Indeed I did the Qt5 port. I tried Qt6 a couple years ago and it was essentially broken, but I'm sure that has been resolved by now.

**Comment by mchamberland** [comment]
Coming back to this two years later, I’d like to actually do the migration now, and I’m looking for a quick sanity check on the approach.

**Where things stand**

- Qt 5.15 standard support ended in May 2025; it’s in extended/commercial-only territory now.
- On macOS, Homebrew has already deprecated qt@5 and will disable it on 19 May 2027.
- On Windows, MSYS2 made Qt6 the default earlier this year (qmake → Qt6; Qt5 is qmake-qt5).
- On Linux, Qt6 packages have been easy to get for a long time (qt6-base-dev, etc.). Distros are actively moving away from Qt5 (Ubuntu’s removal push, Fedora/Arch already Qt6-primary). Qt5 may still be installable on some LTS images, but it’s clearly the legacy path.
- On develop, egs_view already has a fair amount of Qt6 dual-support from [#1164](https://github.com/nrc-cnrc/EGSnrc/pull/1164) (and follow-ups). The other GUIs (egs_gui, egs_inprz, egs_configure) are still Qt5-era. I don’t see an open PR or branch for a full migration.

**Proposal: Qt6-only (drop Qt5)**

When this thread started, dual Qt5/Qt6 support would have been the cautious choice. I don’t think that makes sense anymore. Maintaining #if QT_VERSION forever for an EOL toolkit costs more than it saves, and the remaining “stuck on Qt5” case is mostly locked-down lab images, where IT is usually happier to install a supported package than keep an unsupported one. The Monte Carlo core doesn’t need the GUIs anyway.

So I’d aim to:

1. Branch from current develop (PR against develop).
2. Finish egs_view as Qt6-only (remove the dual #ifdefs rather than extend them).
3. Port egs_gui, egs_inprz, and egs_configure.
4. Update build/docs notes (and the stale Qt4-era bits under HEN_HOUSE/pieces/) for Qt6 on Linux / macOS / Windows.

Happy to do the work. Before I dig in: any objections to Qt6-only, or anything in flight on egs_view I should wait on? @rtownson @ftessier

**Reply by rtownson** [reply]
As far as I know Qt6 is already fully supported in the `develop` branch! Let me know if you spot any issues; I'm happy to have help :)

We are definitely going to keep Qt5 support for a good while longer, I don't see any reason to drop it since it's already done. It's a only a one-command difference in few spots where we check for it, and I haven't come across anything in Qt6 that isn't in Qt5. One reason might be if we were building a brand new fancy GUI where it made a difference to do things the Qt6 way, but as far as I know that isn't planned. In any case, we lean towards supporting old systems and old compilers for as long as possible.

I do have another egs_view update in the works, mostly just improving the performance of egs_editor (which is currently only in `develop`).

**Reply by mchamberland** [reply]
Oh, good to know! I'll give the other apps a try in the develop branch, then. 

Yes, I noticed you do like supporting old tools. I think I saw something about gcc 4.5 fairly recently? 😂

**Reply by mchamberland** [reply]
@rtownson You're right, of course, egs_view on `develop` builds cleanly with Qt6 (minus some warnings).

The other GUI apps under HEN_HOUSE/gui, however, yield the following, and that looks like they haven't been ported to Qt6 yet....?

> ~/Developer/nrc-cnrc/EGSnrc/HEN_HOUSE/gui (qt6-test*) » ls                                                      marc@Moridin
> egs_configure egs_gui       egs_inprz     gui.pro       Makefile
> (base) ----------------------------------------------------------------------------------------------------------------------
> ~/Developer/nrc-cnrc/EGSnrc/HEN_HOUSE/gui (qt6-test*) » make                                                    marc@Moridin
> /opt/homebrew/opt/qt/bin/qmake my_machine=qt6 MAKEFILE=Makefile_qt6 QMAKE=/opt/homebrew/opt/qt/bin/qmake
> Info: creating stash file /Users/marc/Developer/nrc-cnrc/EGSnrc/HEN_HOUSE/gui/.qmake.stash
> /Library/Developer/CommandLineTools/usr/bin/make -f Makefile_qt6
> cd egs_configure/ && ( test -e Makefile || /opt/homebrew/opt/qt/bin/qmake -o Makefile /Users/marc/Developer/nrc-cnrc/EGSnrc/HEN_HOUSE/gui/egs_configure/egs_configure.pro my_machine=qt6 MAKEFILE=Makefile_qt6 QMAKE=/opt/homebrew/opt/qt/bin/qmake ) && /Library/Developer/CommandLineTools/usr/bin/make -f Makefile
> /opt/homebrew/opt/qt/bin/qmake my_machine=qt6 MAKEFILE=Makefile_qt6 the_dir=.
> Project MESSAGE: Dynamic build...
> Project MESSAGE: Dynamic build...
> Project MESSAGE: Dynamic build...
> make -f Makefile_qt6 my_machine=qt6
> /Library/Developer/CommandLineTools/usr/bin/make -f Makefile_qt6.Release
> /Library/Developer/CommandLineTools/usr/bin/clang++ -c -pipe -stdlib=libc++ -O2 -std=gnu++1z  -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -mmacosx-version-min=14.0 -Wall -Wextra -DQT_NO_DEBUG -DQT_WIDGETS_LIB -DQT_GUI_LIB -DQT_XML_LIB -DQT_CORE_LIB -I. -I../egs_gui -I/opt/homebrew/lib/QtWidgets.framework/Headers -I/opt/homebrew/lib/QtGui.framework/Headers -I/opt/homebrew/lib/QtXml.framework/Headers -I/opt/homebrew/lib/QtCore.framework/Headers -I.moc/qt6 -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/OpenGL.framework/Headers -I/opt/homebrew/share/qt/mkspecs/macx-clang -F/opt/homebrew/lib -o .obj/qt6/main.o main.cpp
> In file included from main.cpp:32:
> ./egs_wizard.h:65:52: error: no viable conversion from 'int' to 'Qt::WindowFlags'
>       (aka 'QFlags<WindowType>')
>    65 |   EGS_Wizard(QWidget * parent = 0, Qt::WindowFlags f = 0);
>       |                                                    ^   ~
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:96:33: note: candidate inherited constructor not
>       viable: no known conversion from 'int' to 'QFlag' for 1st argument
>    96 |     constexpr inline Q_IMPLICIT QFlagsStorageHelper(QFlag flag) noexcept
>       |                                 ^                   ~~~~~~~~~~
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:113:17: note: constructor from base class
>       'QFlagsStorageHelper<Qt::WindowType>' inherited here
>   113 |     using Base::Base;
>       |                 ^
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:107:7: note: candidate constructor
>       (the implicit copy constructor) not viable: no known conversion from 'int' to 'const QFlags<Qt::WindowType> &' for 1st
>       argument
>   107 | class QFlags : public QtPrivate::QFlagsStorageHelper<Enum>
>       |       ^~~~~~
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:107:7: note: candidate constructor
>       (the implicit move constructor) not viable: no known conversion from 'int' to 'QFlags<Qt::WindowType> &&' for 1st
>       argument
>   107 | class QFlags : public QtPrivate::QFlagsStorageHelper<Enum>
>       |       ^~~~~~
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:118:33: note: candidate constructor not viable: no
>       known conversion from 'int' to 'Qt::WindowType' for 1st argument
>   118 |     constexpr inline Q_IMPLICIT QFlags(Enum flags) noexcept : Base(std::in_place, Int(flags)) {}
>       |                                 ^      ~~~~~~~~~~
> /opt/homebrew/lib/QtCore.framework/Headers/qflags.h:126:22: note: candidate constructor not viable: no
>       known conversion from 'int' to 'std::initializer_list<Qt::WindowType>' for 1st argument
>   126 |     constexpr inline QFlags(std::initializer_list<Enum> flags) noexcept
>       |                      ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ./egs_wizard.h:65:52: note: passing argument to parameter 'f' here
>    65 |   EGS_Wizard(QWidget * parent = 0, Qt::WindowFlags f = 0);
>       |                                                    ^
> 1 error generated.
> make[4]: *** [.obj/qt6/main.o] Error 1
> make[3]: *** [release] Error 2
> make[2]: *** [all] Error 2
> make[1]: *** [sub-egs_configure-make_first] Error 2
> make: *** [all] Error 2

**Reply by rtownson** [reply]
Thanks for checking this, I mistakenly thought I had done them as well. You are welcome to fix this if you like, otherwise I will likely do it myself soon-ish :)

**Reply by mchamberland** [reply]
I can do it. I'll copy your approach you took with egs_view and make sure Qt5 is still supported.
